core/src/main/java/org/apache/iceberg/Snapshots.java

```diff
@@ -0,0 +1,226 @@
+      return this;
+    }
+
+    public MicroBatchBuilder specsById(Map<Integer, PartitionSpec> specs) {
+      this.specsById = specs;
+      return this;
+    }
+
+    public MicroBatch generate(int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      List<ManifestFile> manifests = isStarting ? snapshot.dataManifests() :
+          snapshot.dataManifests().stream().filter(m -> m.snapshotId().equals(snapshot.snapshotId()))
+              .collect(Collectors.toList());
+
+      List<Pair<ManifestFile, Integer>> manifestIndexes = indexManifests(manifests);
+      List<Pair<ManifestFile, Integer>> skippedManifestIndexes = skipManifests(manifestIndexes, startFileIndex);
+
+      return generateMicroBatch(skippedManifestIndexes, startFileIndex, targetSizeInBytes, isStarting);
+    }
+
+    private List<Pair<ManifestFile, Integer>> indexManifests(List<ManifestFile> manifestFiles) {
+      int currentFileIndex = 0;
+      List<Pair<ManifestFile, Integer>> manifestIndexes = Lists.newArrayList();
+
+      for (ManifestFile manifest : manifestFiles) {
+        int filesCount = manifest.addedFilesCount() + manifest.existingFilesCount();
+        manifestIndexes.add(Pair.of(manifest, currentFileIndex));
+        currentFileIndex += filesCount;
+      }
+
+      return manifestIndexes;
+    }
+
+    private List<Pair<ManifestFile, Integer>> skipManifests(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                                            int startFileIndex) {
+      if (startFileIndex == 0) {
+        return indexedManifests;
+      }
+
+      int index = 0;
+      for (Pair<ManifestFile, Integer> manifest : indexedManifests) {
+        if (manifest.second() > startFileIndex) {
+          break;
+        }
+
+        index++;
+      }
+
+      return indexedManifests.subList(index - 1, indexedManifests.size());
+    }
+
+    private MicroBatch generateMicroBatch(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                          int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      if (indexedManifests.isEmpty()) {
+        return new MicroBatch(snapshot.snapshotId(), startFileIndex, startFileIndex + 1, 0L,
+            CloseableIterable.empty(), true);
+      }
+
+      long currentSizeInBytes = 0L;
+      int currentFileIndex = 0;
+      List<CloseableIterable<FileScanTask>> batchTasks = Lists.newArrayList();
+
+      for (Pair<ManifestFile, Integer> pair : indexedManifests) {
+        currentFileIndex = pair.second();
+
+        CloseableIterable<FileScanTask> tasks;
+        if (isStarting) {
+          tasks = full(ImmutableList.of(pair.first()));
+        } else {
+          tasks = appends(ImmutableList.of(pair.first()));
+        }
+
+        List<FileScanTask> batch = Lists.newArrayList();
+
+        for (FileScanTask task : tasks) {
+          if (currentFileIndex < startFileIndex) {
+            currentFileIndex++;
+            continue;
```

Comment:
Minor: I think it is generally cleaner to have a single invariant to increment in a loop. Using continue causes this to need two, one for each case. Rewriting with an `else` instead of a continue shows it:

```java
if (currentFileIndex < startFileIndex) {
  currentFileIndex++;
} else {
  // add the file to a batch or return
  currentFileIndex++;
}
```

I think it would be better to just reverse the `if` condition and have a single invariant increment:

```java
if (currentFileIndex >= startFileIndex) {
  // add the file to a batch or return
}
currentFileIndex++;
```

Reply:
Changed.